group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/ConcurrentEventQueue.java

```diff
@@ -0,0 +1,258 @@
+    private Condition condition;
+
+    public ConcurrentEventQueue() {
+        this(new Random());
+    }
+
+    public ConcurrentEventQueue(
+        Random random
+    ) {
+        this.random = random;
+        this.queues = new HashMap<>();
+        this.availableKeys = new ArrayList<>();
+        this.inflightKeys = new HashSet<>();
+        this.closed = false;
+        this.lock = new ReentrantLock();
+        this.condition = lock.newCondition();
+    }
+
+    /**
+     * Adds an {{@link Event}} to the queue.
+     *
+     * @param event An {{@link Event}}.
+     */
+    public void add(T event) {
+        lock.lock();
+        try {
+            if (closed) throw new IllegalStateException("Can't accept an event because the queue is close.");
+
+            K key = event.key();
+            Queue<T> queue = queues.get(key);
+            if (queue == null) {
+                queue = new LinkedList<>();
+                queues.put(key, queue);
+                if (!inflightKeys.contains(key)) {
+                    addAvailableKey(key);
+                }
+            }
+            queue.add(event);
+            size++;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Returns the next {{@link Event}} available. This method block indefinitely until
+     * one event is ready or the queue is closed.
+     *
+     * @return The next event.
+     */
+    public T poll() {
+        return poll(Long.MAX_VALUE, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Returns the next {{@link Event}} available. This method blocks for the provided
+     * time and returns null of not event is available.
+     *
+     * @param timeout   The timeout.
+     * @param unit      The timeout unit.
+     * @return The next event available or null.
+     */
+    public T poll(long timeout, TimeUnit unit) {
+        lock.lock();
+        try {
+            K key = randomKey();
+            long nanos = unit.toNanos(timeout);
+            while (key == null && !closed && nanos > 0) {
+                try {
+                    nanos = condition.awaitNanos(nanos);
+                } catch (InterruptedException e) {
+                    // Ignore.
+                }
+                key = randomKey();
+            }
+
+            if (key == null) return null;
+
+            Queue<T> queue = queues.get(key);
+            T event = queue.poll();
+
+            if (queue.isEmpty()) queues.remove(key);
+            inflightKeys.add(key);
+            size--;
+
+            return event;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Marks the event as processed and releases the next event
+     * with the same key. This unblocks waiting threads.
+     *
+     * @param event The event that was processed.
+     */
+    public void done(T event) {
+        lock.lock();
+        try {
+            K key = event.key();
+            inflightKeys.remove(key);
+            if (queues.containsKey(key)) {
```

Comment:
Do we have a method that allows us to remove keys from queues?

Reply:
They are removed [here](https://github.com/apache/kafka/pull/13505/files#diff-bf202b87edf0eb78086b24213f97883a62c57a7a5c966f891d52055aeff01f27R175) in `poll`.

Reply:
That is correct.